lib/bap_types/bap_ir.ml

```diff
@@ -143,18 +143,90 @@ type sub = {
 
 
 type path = int array
-  [@@deriving bin_io, compare, sexp]
-
-
-type program = {
-  subs  : sub term array;
-  paths : path Tid.Table.t;
-} [@@deriving bin_io, fields, sexp]
-
-let compare_program x y =
-  let compare x y = [%compare:sub term array] x y in
-  compare x.subs y.subs
+[@@deriving bin_io, compare, sexp]
+
+
+let mangle_name addr tid name =
+  match addr with
+  | Some a ->
+    sprintf "%s@%s" name @@
+    Bap_bitvector.string_of_value ~hex:true a
+  | None -> sprintf "%s%%%s" name (Tid.to_string tid)
+
+let mangle_sub s =
+  let addr = Dict.find s.dict Bap_attributes.address in
+  let name = mangle_name addr s.tid s.self.name in
+  Tid.set_name s.tid name;
+  let self = {s.self with name} in
+  {s with self}
+
+let names_of_subs subs =
+  let names = String.Table.create () in
+  Array.iter subs ~f:(fun s ->
+      Hashtbl.change names s.self.name ~f:(function
+          | None -> Some [s.tid]
+          | Some tids -> Some (s.tid :: tids)));
+  names
+
+let fix_names ?(old_names=String.Table.create ()) subs =
+  let names = names_of_subs subs in
+  let conflicts = Hashtbl.count names ~f:(fun x -> List.length x > 1) in
+  if conflicts = 0 then subs
+  else
+    let remove name tid = match Hashtbl.find names name with
```

Comment:
instead of using `find`/`set` you can use `change`. Though, this code already is too bad to be discussed at all...